master/internal/kubernetes/pod.go

```diff
@@ -272,6 +273,42 @@ func (p *pod) receivePodStatusUpdate(ctx *actor.Context, msg podStatusUpdate) er
 	return nil
 }
 
+func (p *pod) processMissingPodDeletion(ctx *actor.Context) {
+	ctx.Log().Warn("processing missing pod deletion")
+	if p.container.State == container.Terminated {
+		ctx.Log().Info(
+			"skipping processing missing pod deletion as container is in a terminated state")
+		return
+	}
+
+	// When deleting pods, we do not always receive status updates containing PodFailed
+	// or PodSucceeded. We currently detect this by checking all PodPending messages if
+	// they contain a deletion timestamp.
+	if !p.resourcesDeleted {
+		ctx.Log().Errorf("processing missing pod deletion for a pod that was never deleted")
+	}
+
+	p.container = p.container.Transition(container.Terminated)
+	// Missed pod deletions occur only when a pod is deleted so we assume
+	// that the container was killed.
+	exitCodeConverted := agent.ExitCode(137)
+	containerStopped := agent.ContainerStopped{
+		Failure: &agent.ContainerFailure{
+			FailureType: agent.ContainerFailed,
+			ExitCode:    &exitCodeConverted,
+		},
+	}
+	ctx.Tell(p.taskHandler, sproto.ContainerStateChanged{
```

Comment:
thought: Seems like there could bear to be a helper function for this pair of `Tell`s and `Stop`, which is now appearing four times in this file. (Well, one of them doesn't have the `Stop` because it's called from the `PostStop` handler, but I believe it would be harmless to call `Stop` again.) Separate PR would be fine, though.

Reply:
good idea, added in a helper function, although i did leave the `Stop` call out of it.